File: scratch/baseline.cc
#include "ns3/applications-module.h"
#include "ns3/core-module.h"
#include "ns3/flow-monitor-module.h"
#include "ns3/internet-module.h"
#include "ns3/network-module.h"
#include "ns3/point-to-point-module.h"

#include <fstream>
#include <iomanip>
#include <string>

using namespace ns3;

namespace
{
constexpr uint32_t kIpUdpHeaderBytes = 28;

struct RunMetrics
{
    double throughputMbps{0.0};
    double goodputMbps{0.0};
    double avgDelayMs{0.0};
    double lossRatePct{0.0};
    double overheadRatioPct{0.0};
};

RunMetrics
ComputeMetrics(const FlowMonitor::FlowStats& stats, uint32_t packetSize, double simSeconds)
{
    RunMetrics m;

    if (simSeconds > 0.0)
    {
        m.throughputMbps = static_cast<double>(stats.rxBytes) * 8.0 / simSeconds / 1e6;
        m.goodputMbps = static_cast<double>(stats.rxPackets) * packetSize * 8.0 / simSeconds / 1e6;
    }

    if (stats.rxPackets > 0)
    {
        m.avgDelayMs = stats.delaySum.GetSeconds() * 1000.0 / static_cast<double>(stats.rxPackets);
        const double totalBytes = static_cast<double>(stats.rxPackets) * (packetSize + kIpUdpHeaderBytes);
        const double headerBytes = static_cast<double>(stats.rxPackets) * kIpUdpHeaderBytes;
        m.overheadRatioPct = (totalBytes > 0.0) ? (headerBytes / totalBytes) * 100.0 : 0.0;
    }

    if (stats.txPackets > 0)
    {
        m.lossRatePct = static_cast<double>(stats.txPackets - stats.rxPackets) * 100.0 /
                        static_cast<double>(stats.txPackets);
    }

    return m;
}
} // namespace

int
main(int argc, char* argv[])
{
    uint32_t packetSize = 512;
    double simSeconds = 10.0;
    std::string outputCsv = "project/results/baseline.csv";

    CommandLine cmd(__FILE__);
    cmd.AddValue("packetSize", "Application payload size in bytes", packetSize);
    cmd.AddValue("simSeconds", "Simulation duration in seconds", simSeconds);
    cmd.AddValue("outputCsv", "Output CSV file path", outputCsv);
    cmd.Parse(argc, argv);

    RngSeedManager::SetSeed(206);
    RngSeedManager::SetRun(1);

    NodeContainer nodes;
    nodes.Create(2);

    PointToPointHelper p2p;
    p2p.SetDeviceAttribute("DataRate", StringValue("10Mbps"));
    p2p.SetChannelAttribute("Delay", StringValue("2ms"));

    NetDeviceContainer devices = p2p.Install(nodes);

    InternetStackHelper internet;
    internet.Install(nodes);

    Ipv4AddressHelper ipv4;
    ipv4.SetBase("10.1.1.0", "255.255.255.0");
    Ipv4InterfaceContainer interfaces = ipv4.Assign(devices);

    uint16_t port = 9000;
    PacketSinkHelper sinkHelper("ns3::UdpSocketFactory",
                                InetSocketAddress(Ipv4Address::GetAny(), port));
    ApplicationContainer sinkApp = sinkHelper.Install(nodes.Get(1));
    sinkApp.Start(Seconds(0.0));
    sinkApp.Stop(Seconds(simSeconds + 0.5));

    OnOffHelper onoff("ns3::UdpSocketFactory", Address(InetSocketAddress(interfaces.GetAddress(1), port)));
    onoff.SetAttribute("DataRate", DataRateValue(DataRate("8Mbps")));
    onoff.SetAttribute("PacketSize", UintegerValue(packetSize));
    onoff.SetAttribute("OnTime", StringValue("ns3::ConstantRandomVariable[Constant=1]"));
    onoff.SetAttribute("OffTime", StringValue("ns3::ConstantRandomVariable[Constant=0]"));

    ApplicationContainer senderApp = onoff.Install(nodes.Get(0));
    senderApp.Start(Seconds(1.0));
    senderApp.Stop(Seconds(simSeconds));

    FlowMonitorHelper flowHelper;
    Ptr<FlowMonitor> monitor = flowHelper.InstallAll();

    Simulator::Stop(Seconds(simSeconds + 1.0));
    Simulator::Run();

    monitor->CheckForLostPackets();
    Ptr<Ipv4FlowClassifier> classifier = DynamicCast<Ipv4FlowClassifier>(flowHelper.GetClassifier());
    const auto statsMap = monitor->GetFlowStats();

    FlowMonitor::FlowStats flowStats;
    bool foundFlow = false;

    for (const auto& kv : statsMap)
    {
        Ipv4FlowClassifier::FiveTuple t = classifier->FindFlow(kv.first);
        if (t.destinationPort == port)
        {
            flowStats = kv.second;
            foundFlow = true;
            break;
        }
    }

    Simulator::Destroy();

    if (!foundFlow)
    {
        NS_LOG_UNCOND("No matching flow was found in FlowMonitor statistics.");
        return 1;
    }

    RunMetrics m = ComputeMetrics(flowStats, packetSize, simSeconds - 1.0);

    std::ofstream out(outputCsv, std::ios::trunc);
    if (!out)
    {
        NS_LOG_UNCOND("Failed to open output CSV: " << outputCsv);
        return 1;
    }

    out << "scenario,bandwidth_mbps,packet_size_bytes,throughput_mbps,goodput_mbps,avg_delay_ms,"
           "packet_loss_rate_pct,overhead_ratio_pct\n";
    out << std::fixed << std::setprecision(6);
    out << "baseline,10," << packetSize << ',' << m.throughputMbps << ',' << m.goodputMbps << ','
        << m.avgDelayMs << ',' << m.lossRatePct << ',' << m.overheadRatioPct << '\n';

    NS_LOG_UNCOND("Baseline results written to: " << outputCsv);
    return 0;
}

File: projectmain/simulations_py/baseline.py
#!/usr/bin/env python3
import csv
from ns import ns

IP_UDP_HEADER_BYTES = 28


def compute_metrics(stats, packet_size, active_time):
    throughput = (stats.rxBytes * 8.0) / active_time / 1e6 if active_time > 0 else 0.0
    goodput = (stats.rxPackets * packet_size * 8.0) / active_time / 1e6 if active_time > 0 else 0.0
    delay_ms = (stats.delaySum.GetSeconds() * 1000.0 / stats.rxPackets) if stats.rxPackets > 0 else 0.0
    loss_pct = ((stats.txPackets - stats.rxPackets) * 100.0 / stats.txPackets) if stats.txPackets > 0 else 0.0

    total_bytes = stats.rxPackets * (packet_size + IP_UDP_HEADER_BYTES)
    header_bytes = stats.rxPackets * IP_UDP_HEADER_BYTES
    overhead_pct = (header_bytes * 100.0 / total_bytes) if total_bytes > 0 else 0.0
    return throughput, goodput, delay_ms, loss_pct, overhead_pct


def main():
    packet_size = 512
    sim_seconds = 10.0
    output_csv = "project/results/baseline_py.csv"
    port = 9000

    ns.RngSeedManager.SetSeed(206)
    ns.RngSeedManager.SetRun(1)

    nodes = ns.NodeContainer()
    nodes.Create(2)

    p2p = ns.PointToPointHelper()
    p2p.SetDeviceAttribute("DataRate", ns.StringValue("10Mbps"))
    p2p.SetChannelAttribute("Delay", ns.StringValue("2ms"))
    devices = p2p.Install(nodes)

    internet = ns.InternetStackHelper()
    internet.Install(nodes)

    ipv4 = ns.Ipv4AddressHelper()
    ipv4.SetBase(ns.Ipv4Address("10.1.1.0"), ns.Ipv4Mask("255.255.255.0"))
    interfaces = ipv4.Assign(devices)

    sink = ns.PacketSinkHelper(
        "ns3::UdpSocketFactory",
        ns.InetSocketAddress(ns.Ipv4Address.GetAny(), port).ConvertTo(),
    )
    sink_apps = sink.Install(ns.NodeContainer(nodes.Get(1)))
    sink_apps.Start(ns.Seconds(0.0))
    sink_apps.Stop(ns.Seconds(sim_seconds + 0.5))

    remote = ns.InetSocketAddress(interfaces.GetAddress(1), port)
    onoff = ns.OnOffHelper("ns3::UdpSocketFactory", remote.ConvertTo())
    onoff.SetAttribute("DataRate", ns.DataRateValue(ns.DataRate("8Mbps")))
    onoff.SetAttribute("PacketSize", ns.UintegerValue(packet_size))
    onoff.SetAttribute("OnTime", ns.StringValue("ns3::ConstantRandomVariable[Constant=1]"))
    onoff.SetAttribute("OffTime", ns.StringValue("ns3::ConstantRandomVariable[Constant=0]"))

    sender_apps = onoff.Install(ns.NodeContainer(nodes.Get(0)))
    sender_apps.Start(ns.Seconds(1.0))
    sender_apps.Stop(ns.Seconds(sim_seconds))

    flow = ns.FlowMonitorHelper()
    monitor = flow.InstallAll()

    ns.Simulator.Stop(ns.Seconds(sim_seconds + 1.0))
    ns.Simulator.Run()
    monitor.CheckForLostPackets()

    classifier = flow.GetClassifier()
    picked = None
    for flow_id, st in monitor.GetFlowStats():
        t = classifier.FindFlow(flow_id)
        if t.destinationPort == port:
            picked = st
            break

    ns.Simulator.Destroy()

    if picked is None:
        raise RuntimeError("No matching flow found")

    m = compute_metrics(picked, packet_size, sim_seconds - 1.0)

    with open(output_csv, "w", newline="", encoding="utf-8") as f:
        w = csv.writer(f)
        w.writerow([
            "scenario",
            "bandwidth_mbps",
            "packet_size_bytes",
            "throughput_mbps",
            "goodput_mbps",
            "avg_delay_ms",
            "packet_loss_rate_pct",
            "overhead_ratio_pct",
        ])
        w.writerow([
            "baseline",
            10,
            packet_size,
            f"{m[0]:.6f}",
            f"{m[1]:.6f}",
            f"{m[2]:.6f}",
            f"{m[3]:.6f}",
            f"{m[4]:.6f}",
        ])

    print(f"Baseline results written to: {output_csv}")


if __name__ == "__main__":
    main()
